lib/sycamore/sycamore/transforms/base.py

```diff
@@ -0,0 +1,127 @@
+import logging
+from typing import Any, Callable, Iterable, Optional
+
+import numpy as np
+from ray.data import ActorPoolStrategy, Dataset
+
+from sycamore.data import Document, MetadataDocument
+from sycamore.plan_nodes import Node, UnaryNode
+
+
+def _noneOr(a: Any, default: Any) -> Any:
+    if a is None:
+        return default
+    return a
+
+
+def rename(new_function_name: str):
+    def decorator(f):
+        f.__name__ = new_function_name
+        return f
+
+    return decorator
+
+
+class BaseMap(UnaryNode):
+    """
+    BaseMap abstracts away MetadataDocuments from all other transforms.
+    """
+
+    def __init__(
+        self,
+        child: Node,
+        *,
+        f: Callable[[list[Document]], list[Document]],
+        name: Optional[str] = None,
+        args: Optional[Iterable[Any]] = None,
+        kwargs: Optional[dict[str, Any]] = None,
+        constructor_args: Optional[Iterable[Any]] = None,
+        constructor_kwargs: Optional[dict[str, Any]] = None,
+        **resource_args,
+    ):
+        super().__init__(child, **resource_args)
+        if name is None:
+            if isinstance(f, type):
+                name = type(f).__name__
+            else:
+                name = f.__name__
+
+        self._f = f
+        self._name = name
+        self._args = args
+        self._kwargs = kwargs
+        self._constructor_args = constructor_args
+        self._constructor_kwargs = constructor_kwargs
+
+    def execute(self) -> "Dataset":
+        input_dataset = self.child().execute()
+
+        if isinstance(self._f, type):
+            return input_dataset.map_batches(self._map_class(), compute=ActorPoolStrategy(size=1), **self.resource_args)
+        else:
+            return input_dataset.map_batches(self._map_function(), **self.resource_args)
+
+    def _map_function(self):
+        f = self._f
+        name = self._name
+        args = _noneOr(self._args, tuple())
+        kwargs = _noneOr(self._kwargs, {})
+
+        @rename(name)
+        def ray_callable(ray_input: dict[str, np.ndarray]) -> dict[str, list]:
+            return BaseMap._process_ray(ray_input, name, lambda d: f(d, *args, **kwargs))
+
+        return ray_callable
+
+    def _map_class(self):
```

Comment:
One thing I'm trying to understand is whether this has any impact on when Ray will create an actor or a task? At the moment I believe we use an actor when you try to map a class (which I think is from [here](https://github.com/aryn-ai/sycamore/blob/13322f80ec3610063b1073b14b1d006f44d0e8c1/lib/sycamore/sycamore/transforms/map.py#L37)). I don't know that we have to do it exactly the way we do it now, but would be good to understand if there are changes.

Reply:
It is not intended to have an impact. In class mode it makes a single instance and calls it repeatedly (as verified by test). I think that's what the old code also does.